File: zatial funkcne api a lib/AeroShield.cpp
/*
  API for the AeroShield hardware.
  
  The file is a part of the application programming interface for
  the AeroShield didactic tool for control engineering and 
  mechatronics education.
  
  This code is part of the AutomationShield hardware and software
  ecosystem. Visit http://www.automationshield.com for more
  details. This code is licensed under a Creative Commons
  Attribution-NonCommercial 4.0 International License.

  Created by .... 
  Last update: 05.02.2022.
*/


#include "AeroShield.h"



// Initializes hardware pins
void AeroShieldClass::begin(){  
      pinMode(AERO_UPIN,OUTPUT);  		// Thermistor pin
      Wire.begin();                       //
}


float AeroShieldClass::convertRawAngleToDegrees(word newAngle) {
  float retVal = newAngle * 0.087;
  ang = retVal;
  return ang;
}


float AeroShieldClass::calibration(word RawAngle) {

  AutomationShield.serialPrint("Calibration running...\n");
  startangle=0;
  analogWrite(AERO_UPIN,50);
  delay(250);
  analogWrite(AERO_UPIN,0);
  delay(5000);
  
  startangle = RawAngle;
  analogWrite(AERO_UPIN,0);
    for(int i=0;i<3;i++){
      analogWrite(AERO_UPIN,1);
      delay(200);
      analogWrite(AERO_UPIN,0);
      delay(200);
      }

  AutomationShield.serialPrint("Calibration done");
    return startangle;
    /*     
  while(angle<=startangle+90.0){
    analogWrite(AERO_UPIN,MOTOR_PWM);
    delay(300);
    angle=AeroShield.convertRawAngleToDegrees(ams5600.getRawAngle());
    MOTOR_PWM=MOTOR_PWM+5;
    }
    */
  }

/*
void AeroShieldClass::actuatorWrite(float percent){  
      analogWrite(HEAT_UPIN,AutomationShield.percToPwm(percent)); // Write to actuator
}

/*
// Reads thermistor voltage
float AeroShieldClass::getThermistorVoltage() { 
      return (float)analogRead(HEAT_YPIN) * ARES;
}

// Reads thermistor resistance
// Voltage divider circuit (Kirchhoffs law)
float AeroShieldClass::getThermistorResistance() {
	float Vterm = getThermistorVoltage();
	 return ((Vterm*VD_RES)/(VD_REF-Vterm));
}*/

AeroShieldClass AeroShield;

File: MATLAB/AeroShield.m
%   AeroShield MATLAB API.
%
%
%   This code is part of the AutomationShield hardware and software
%   ecosystem. Visit http://www.automationshield.com for more
%   details. This code is licensed under a Creative Commons
%   Attribution-NonCommercial 4.0 International License.
%
%   Created by Peter Tibensky.
%   Last update: 13.3.2022.

classdef AeroShield < handle

    properties
        arduino;
        as5600;
    end

    properties(Constant)
            AERO_UPIN = 'D5';          % actuator pin
            AERO_RPIN = 'A3';          % potentiometer pin
            VOLTAGE_SENSOR_PIN = 'A2'; % current measure pin 
            voltageReference = 5.0;    % voltage reference value
            ShuntRes = 0.1;            % value of shunt resistor in ohms
            correction1 = 4.1220;	   % correction for measuring current 				  
            correction2 = 0.33;        % correction for measuring current
            repeatTimes = 50;          % number of loops for current measure
    end

    methods
        function begin(AeroShieldObject)          % initialization function
            AeroShieldObject.arduino = arduino(); % constructing arduino object 
            AeroShieldObject.as5600 = device(AeroShieldObject.arduino,'I2CAddress',0x36); % constructing sensor object
            configurePin(AeroShieldObject.arduino,AeroShieldObject.AERO_UPIN, 'DigitalOutput') % pin configuration as output
            disp('AeroShield initialized.') % display message 
        end
        
         function startangle = calibration(AeroShieldObject) % calibration function 
            write(AeroShieldObject.as5600, 0x0c, 'uint8');  % send command for reading angle
            write(AeroShieldObject.as5600, 0x0d, 'uint8');  % send command for reading angle
            startangle = read(AeroShieldObject.as5600, 1, 'uint16'); % read answer from sensor
        end

        function PWM = referenceRead(AeroShieldObject)  % reference read 
           PWM= readVoltage(AeroShieldObject.arduino, AeroShieldObject.AERO_RPIN); % read potentiometer value 
        end
        
        function actuatorWrite(AeroShieldObject, PWM) % actuator write 
            writePWMVoltage(AeroShieldObject.arduino, AeroShieldObject.AERO_UPIN, PWM); % write voltage level to the motor pin
        end
       
        
        function RAW = getRawAngle(AeroShieldObject)              % get raw angle value 
            write(AeroShieldObject.as5600, 0x0c, 'uint8');        % send command for reading angle
            write(AeroShieldObject.as5600, 0x0d, 'uint8');        % send command for reading angle
            RAW = read(AeroShieldObject.as5600, 1, 'uint16');     % read answer from sensor
        end
        
        
        function currentMean = getCurrent()  % get current reading 
              for r = 1:repeatTimes
              voltageValue = readVoltage(AeroShieldObject.arduino, AeroShieldObject.VOLTAGE_SENSOR_PIN); % read pin voltage value 
              voltageValue= (voltageValue * voltageReference) / 1024;                      % Remap the ADC value into a voltage number (5V reference)
              Current= Current + correction1-(voltageValue / (10 * ShuntRes));             % Equation given by the INA169 datasheet to
              end                                                                          % determine the current flowing through ShuntRes. RL = 10k          
                                                                                           % Is = (Vout x 1k) / (RS x RL)
        currentMean= Current/repeatTimes;                                                  % Callculating mean current value 
        currentMean= currentMean-correction2;                                              % Small correction of current value(determined by multimeter)
        if currentMean < 0.000                                                             % Correction for occasional bug causing the value to be negative. 
        currentMean= 0.000;                                                                % When it so happens, zero out the value. 
        end
        current= 0;                                                                        % Zero out current value        
        voltageValue= 0;                                                                                                                                                                          
        end 
        
    end
end